### PR TITLE
[Kandinsky 3.0] Follow-up TODOs

### PR DESCRIPTION
**Pull request by yiyixuxu**
work through the remaining TODOs from https://github.com/huggingface/diffusers/pull/5913
- [x] Treat all the TODO statemens in the code
- [x] Add better docs (https://github.com/huggingface/diffusers/pull/5937#event-11086056097)
- [x] Add tests for img2img
- [x] Rename the pipeline and model files
- [ ] Clean up all the unet blocks and get rid of hard to read code
- [ ] Publish on discord etc...

#### text-2-image
```python
from diffusers import AutoPipelineForText2Image
import torch

pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-3", variant="fp16", torch_dtype=torch.float16)
pipe.enable_model_cpu_offload()
        
prompt = "A photograph of the inside of a subway train. There are raccoons sitting on the seats. One of them is reading a newspaper. The window shows the city in the background."

generator = torch.Generator(device="cpu").manual_seed(0)
image = pipe(prompt, num_inference_steps=25, generator=generator).images[0]
```
yiyi_test_1_out

#### image-2-image

```python
from diffusers import AutoPipelineForImage2Image
from diffusers.utils import load_image
import torch

pipe = AutoPipelineForImage2Image.from_pretrained("kandinsky-community/kandinsky-3", variant="fp16", torch_dtype=torch.float16)
pipe.enable_model_cpu_offload()
        
prompt = "A painting of the inside of a subway train with tiny raccoons."
image = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/kandinsky3/t2i.png")

generator = torch.Generator(device="cpu").manual_seed(0)
image = pipe(prompt, image=image, strength=0.75, num_inference_steps=25, generator=generator).images[0]
```
yiyi_test_2_out

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -593,7 +597,7 @@ def get_attention_scores(
             beta = 0
         else:
             baddbmm_input = attention_mask
-            beta = 1
+            beta = self.scale_mask_factor
```
Comment (yiyixuxu) [inline src/diffusers/models/attention_processor.py:600]: added a new config for `Attention` here

set this to be a large negative number helps a lot with numerical stability. in kandinsky they "fill" the empty tokens 
in attention_matrix with largest possible negative number(see code -> 
https://github.com/huggingface/diffusers/blob/5ae3c3a56b52094cd01d184da38255136d7715e7/src/diffusers/models/attention_processor.py#L2251)

I set this config to be `-6000` for simplicity - not exactly same but seem to be sufficient.

File: src/diffusers/models/attention_processor.py [hunk 2]
```diff
@@ -109,15 +109,19 @@ def __init__(
         residual_connection: bool = False,
         _from_deprecated_attn_block: bool = False,
         processor: Optional["AttnProcessor"] = None,
+        scale_mask_factor: float = 1.0,
+        out_dim: int = None,
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:113]: Is `out_dim` different from `query_dim` here?

File: src/diffusers/models/attention_processor.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/models/attention_processor.py:600]: Hmm is `beta` supposed to be used to control mask precision?

File: src/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -2219,44 +2223,6 @@ def __call__(
         return hidden_states
 
 
-# TODO(Yiyi): This class should not exist, we can replace it with a normal attention processor I believe
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:2222]: Nice!

File: src/diffusers/models/unet_2d_condition_kandi3.py [hunk 4]
```diff
@@ -1,16 +1,28 @@
-import math
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition_kandi3.py:1]: Actually can we rename this file to `unet_kandinsky3.py` ? I don't like `kandi..` much

File: src/diffusers/models/unet_2d_condition_kandi3.py [hunk 5]
```diff
@@ -572,14 +520,10 @@ def forward(self, x, time_embed, context=None, context_mask=None, image_mask=Non
...
-        if image_mask is not None:
-            mask_height, mask_width = image_mask.shape[-2:]
-            kernel_size = (mask_height // height, mask_width // width)
-            image_mask = F.max_pool2d(image_mask, kernel_size, kernel_size)
-            image_mask = image_mask.reshape(image_mask.shape[0], -1)
-
-        out = self.attention(out, context, context_mask, image_mask)
+        out = self.attention(out, context, context_mask)
         out = out.permute(0, 2, 1).unsqueeze(-1).reshape(out.shape[0], -1, height, width)
         x = x + out
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition_kandi3.py:528]: very nice clean-ups!

File: src/diffusers/models/attention_processor.py (hunk 2 above)
Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py:113]: the only difference is the `to_out`layer here - Kandinsky attention output does not change dimension from `inner_dim`, while our attention class will project the output to `query_dim`. I added an `out_dim` for this purpose, but I think we can certainly add a different config to do this if it makes more sense! 

        self.to_out.append(nn.Linear(out_channels, out_channels, bias=False))`

https://github.com/huggingface/diffusers/blob/d1b2a1a95710964ff84b2c2332cd2ad93734561e/src/diffusers/models/unet_kandi3.py#L453

File: src/diffusers/models/attention_processor.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py:600]: actually, I think I should do this instead! https://github.com/huggingface/diffusers/blob/d1b2a1a95710964ff84b2c2332cd2ad93734561e/src/diffusers/models/unet_2d_condition.py#L896

This `beta` here is essentially trying to do the same thing - it's trying to zero out the zero token's attention score in the softmax operation. I did not realize I was missing this step because Kandinsky cuts off (most except for one) the zero tokens from `prompt_embeds`, so not doing this step still generates accurate output for the most part

Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py]: refactored and now this script (one scenario when the attention_mask actually need to be applied) not exactly but similar outputs on main and branch 

```python
from diffusers import AutoPipelineForText2Image
import torch

pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-3", variant="fp16", torch_dtype=torch.float16)
pipe.enable_model_cpu_offload()
        
prompt = ["A photograph of the inside of a subway train. There are raccoons sitting on the seats. One of them is reading a newspaper. The window shows the city in the background.",
          "A photograph of the inside of a subway train. There are raccoons sitting on the seats. One of them is reading a newspaper. The window shows the city in the background. A photograph of the inside of a subway train. There are raccoons sitting on the seats. One of them is reading a newspaper. The window shows the city in the background."]

generator = [torch.Generator(device="cpu").manual_seed(0),torch.Generator(device="cpu").manual_seed(1)]
image = pipe(prompt, num_inference_steps=25, generator=generator).images[0]
```

main         |  branch
:-------------------------:|:-------------------------:
yiyi_test_3_out_bm | yiyi_test_3_out

**Comment by patrickvonplaten** [comment]
@yiyixuxu lemme know once ready for a final review :-)

File: src/diffusers/models/attention_processor.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/models/attention_processor.py:112]: That works! Makes sense

**Review by patrickvonplaten — APPROVED** [review]
Great clean-up - thanks!
